/*
* ProfaneDB - A Protocol Buffers database.
* Copyright (C) 2017  "Giorgio Azzinnaro" <giorgio.azzinnaro@gmail.com>
*
* it under the terms of the GNU General Public License as published by
* This program is free software: you can redistribute it and/or modify
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/


syntax = "proto3";

package org.ltc.iltalk.core.protobuf.jpdb;

option go_package = "gitlab.com/profanedb/protobuf/db";
option csharp_namespace = "ProfaneDB.Protobuf";
option java_package = "com.profanedb.protobuf";
option objc_class_prefix = "PDB";

import "google/protobuf/any.proto";

import "org/ltc/iltalk/core/protobuf/jpdb/storage.proto";

service Db {
    rpc Get (GetReq) returns (GetResp);

    rpc Put (PutReq) returns (PutResp);

    rpc Delete (DelReq) returns (DelResp);
}

message GetReq {
    Key key = 1; // TODO How about getting multiple objects at once
}

message GetResp {
    google.protobuf.Any message = 1;
}

message PutReq {
    google.protobuf.Any serializable = 1;
}

message PutResp {
    Key key = 1; // TODO Maybe should be repeated for all stored keys?
}

message DelReq {
    Key key = 1;
}

message DelResp {
}
